CoreServices/WindowServer/libinput/tools/libinput-measure-touchpad-size.man
.TH libinput-measure-touchpad-size "1"
.SH NAME
libinput\-measure\-touchpad\-size \- measure the size of a touchpad
.SH SYNOPSIS
.B libinput measure touchpad\-size [\-\-help] WxH [\fI/dev/input/event0\fI]
.SH DESCRIPTION
.PP
The
.B "libinput measure touchpad\-size"
tool measures the size of a touchpad. This is an interactive tool. When
executed, the tool will prompt the user to interact with the touchpad. The
tool records the axis ranges and calculates the size and resolution based on
those. On termination, the tool prints a hwdb entry that can be added to the
.B 60-evdev.hwdb
file provided by system.
.PP
For details see the online documentation here:
.I https://wayland.freedesktop.org/libinput/doc/latest/absolute-coordinate-ranges.html
.PP
This is a debugging tool only, its output may change at any time. Do not
rely on the output.
.PP
This tool usually needs to be run as root to have access to the
/dev/input/eventX nodes.
.SH OPTIONS
This tool must be provided the physical dimensions of the device in mm.
For example, if your touchpad is 100mm wide and 55mm high, run this tool as
.B libinput measure touchpad-size 100x55
.PP
If a device node is given, this tool opens that device node. Otherwise, this
tool searches for the first node that looks like a touchpad and uses that
node.
.TP 8
.B \-\-help
Print help
.SH LIBINPUT
Part of the
.B libinput(1)
suite


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

CoreServices/WindowServer/libinput/tools/libinput-measure-touchpad-size.py
#!/usr/bin/env python3
# vim: set expandtab shiftwidth=4:
# -*- Mode: python; coding: utf-8; indent-tabs-mode: nil -*- */
#
# Copyright © 2020 Red Hat, Inc.
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice (including the next
# paragraph) shall be included in all copies or substantial portions of the
# Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.


import sys
import argparse

try:
    import libevdev
    import pyudev
except ModuleNotFoundError as e:
    print("Error: {}".format(str(e)), file=sys.stderr)
    print(
        "One or more python modules are missing. Please install those "
        "modules and re-run this tool."
    )
    sys.exit(1)


class DeviceError(Exception):
    pass


class Point:
    def __init__(self, x=None, y=None):
        self.x = x
        self.y = y


class Touchpad(object):
    def __init__(self, evdev):
        x = evdev.absinfo[libevdev.EV_ABS.ABS_X]
        y = evdev.absinfo[libevdev.EV_ABS.ABS_Y]
        if not x or not y:
            raise DeviceError("Device does not have an x or axis")

        if not x.resolution or not y.resolution:
            print("Device does not have resolutions.", file=sys.stderr)
            x.resolution = 1
            y.resolution = 1

        self.xrange = x.maximum - x.minimum
        self.yrange = y.maximum - y.minimum
        self.width = self.xrange / x.resolution
        self.height = self.yrange / y.resolution

        self._x = x
        self._y = y

        # We try to make the touchpad at least look proportional. The
        # terminal character space is (guesswork) ca 2.3 times as high as
        # wide.
        self.columns = 30
        self.rows = int(
            self.columns
            * (self.yrange // y.resolution)
            // (self.xrange // x.resolution)
            / 2.3
        )
        self.pos = Point(0, 0)
        self.min = Point()
        self.max = Point()

    @property
    def x(self):
        return self._x

    @property
    def y(self):
        return self._y

    @x.setter
    def x(self, x):
        self._x.minimum = min(self.x.minimum, x)
        self._x.maximum = max(self.x.maximum, x)
        self.min.x = min(x, self.min.x or 0xFFFFFFFF)
        self.max.x = max(x, self.max.x or -0xFFFFFFFF)
        # we calculate the position based on the original range.
        # this means on devices with a narrower range than advertised, not
        # all corners may be reachable in the touchpad drawing.
        self.pos.x = min(0.99, (x - self._x.minimum) / self.xrange)

    @y.setter
    def y(self, y):
        self._y.minimum = min(self.y.minimum, y)
        self._y.maximum = max(self.y.maximum, y)
        self.min.y = min(y, self.min.y or 0xFFFFFFFF)
        self.max.y = max(y, self.max.y or -0xFFFFFFFF)
        # we calculate the position based on the original range.
        # this means on devices with a narrower range than advertised, not
        # all corners may be reachable in the touchpad drawing.
        self.pos.y = min(0.99, (y - self._y.minimum) / self.yrange)

    def update_from_data(self):
        if None in [self.min.x, self.min.y, self.max.x, self.max.y]:
            raise DeviceError("Insufficient data to continue")
        self._x.minimum = self.min.x
        self._x.maximum = self.max.x
        self._y.minimum = self.min.y
        self._y.maximum = self.max.y

    def draw(self):
        print(
            "Detected axis range: x [{:4d}..{:4d}], y [{:4d}..{:4d}]".format(
                self.min.x if self.min.x is not None else 0,
                self.max.x if self.max.x is not None else 0,
                self.min.y if self.min.y is not None else 0,
                self.max.y if self.max.y is not None else 0,
            )
        )

        print()
        print("Move one finger along all edges of the touchpad".center(self.columns))
        print("until the detected axis range stops changing.".center(self.columns))

        top = int(self.pos.y * self.rows)

        print("+{}+".format("".ljust(self.columns, "-")))
        for row in range(0, top):
            print("|{}|".format("".ljust(self.columns)))

        left = int(self.pos.x * self.columns)
        right = max(0, self.columns - 1 - left)
        print("|{}{}{}|".format("".ljust(left), "O", "".ljust(right)))

        for row in range(top + 1, self.rows):
            print("|{}|".format("".ljust(self.columns)))

        print("+{}+".format("".ljust(self.columns, "-")))

        print("Press Ctrl+C to stop".center(self.columns))

        print("\033[{}A".format(self.rows + 8), flush=True)

        self.rows_printed = self.rows + 8

    def erase(self):
        # Erase all previous lines so we're not left with rubbish
        for row in range(self.rows_printed):
            print("\033[K")
        print("\033[{}A".format(self.rows_printed))


def dimension(string):
    try:
        ts = string.split("x")
        t = tuple([int(x) for x in ts])
        if len(t) == 2:
            return t
    except:  # noqa
        pass

    msg = "{} is not in format WxH".format(string)
    raise argparse.ArgumentTypeError(msg)


def between(v1, v2, deviation):
    return v1 - deviation < v2 < v1 + deviation


def dmi_modalias_match(modalias):
    modalias = modalias.split(":")
    dmi = {"svn": None, "pvr": None, "pn": None}
    for m in modalias:
        for key in dmi:
            if m.startswith(key):
                dmi[key] = m[len(key) :]

    # Based on the current 60-evdev.hwdb, Lenovo uses pvr and everyone else
    # uses pn to provide a human-identifiable match
    if dmi["svn"] == "LENOVO":
        return "dmi:*svn{}:*pvr{}*".format(dmi["svn"], dmi["pvr"])
    else:
        return "dmi:*svn{}:*pn{}*".format(dmi["svn"], dmi["pn"])


def main(args):
    parser = argparse.ArgumentParser(description="Measure the touchpad size")
    parser.add_argument(
        "size",
        metavar="WxH",
        type=dimension,
        help="Touchpad size (width by height) in mm",
    )
    parser.add_argument(
        "path",
        metavar="/dev/input/event0",
        nargs="?",
        type=str,
        help="Path to device (optional)",
    )
    context = pyudev.Context()

    args = parser.parse_args()
    if not args.path:
        for device in context.list_devices(subsystem="input"):
            if device.get("ID_INPUT_TOUCHPAD", 0) and (
                device.device_node or ""
            ).startswith("/dev/input/event"):
                args.path = device.device_node
                name = "unknown"
                parent = device
                while parent is not None:
                    n = parent.get("NAME", None)
                    if n:
                        name = n
                        break
                    parent = parent.parent

                print("Using {}: {}".format(name, device.device_node))
                break
        else:
            print("Unable to find a touchpad device.", file=sys.stderr)
            return 1

    dev = pyudev.Devices.from_device_file(context, args.path)
    overrides = [p for p in dev.properties if p.startswith("EVDEV_ABS")]
    if overrides:
        print()
        print("********************************************************************")
        print("WARNING: axis overrides already in place for this device:")
        for prop in overrides:
            print("  {}={}".format(prop, dev.properties[prop]))
        print("The systemd hwdb already overrides the axis ranges and/or resolution.")
        print("This tool is not needed unless you want to verify the axis overrides.")
        print("********************************************************************")
        print()

    try:
        fd = open(args.path, "rb")
        evdev = libevdev.Device(fd)
        touchpad = Touchpad(evdev)
        print(
            "Kernel specified touchpad size: {:.1f}x{:.1f}mm".format(
                touchpad.width, touchpad.height
            )
        )
        print("User specified touchpad size:   {:.1f}x{:.1f}mm".format(*args.size))

        print()
        print(
            "Kernel axis range:   x [{:4d}..{:4d}], y [{:4d}..{:4d}]".format(
                touchpad.x.minimum,
                touchpad.x.maximum,
                touchpad.y.minimum,
                touchpad.y.maximum,
            )
        )

        print("Put your finger on the touchpad to start\033[1A")

        try:
            touchpad.draw()
            while True:
                for event in evdev.events():
                    if event.matches(libevdev.EV_ABS.ABS_X):
                        touchpad.x = event.value
                    elif event.matches(libevdev.EV_ABS.ABS_Y):
                        touchpad.y = event.value
                    elif event.matches(libevdev.EV_SYN.SYN_REPORT):
                        touchpad.draw()
        except KeyboardInterrupt:
            touchpad.erase()
            touchpad.update_from_data()

        print(
            "Detected axis range: x [{:4d}..{:4d}], y [{:4d}..{:4d}]".format(
                touchpad.x.minimum,
                touchpad.x.maximum,
                touchpad.y.minimum,
                touchpad.y.maximum,
            )
        )

        touchpad.x.resolution = round(
            (touchpad.x.maximum - touchpad.x.minimum) / args.size[0]
        )
        touchpad.y.resolution = round(
            (touchpad.y.maximum - touchpad.y.minimum) / args.size[1]
        )

        print(
            "Resolutions calculated based on user-specified size: x {}, y {} units/mm".format(
                touchpad.x.resolution, touchpad.y.resolution
            )
        )

        # If both x/y are within some acceptable deviation, we skip the axis
        # overrides and only override the resolution
        xorig = evdev.absinfo[libevdev.EV_ABS.ABS_X]
        yorig = evdev.absinfo[libevdev.EV_ABS.ABS_Y]
        deviation = 1.5 * touchpad.x.resolution  # 1.5 mm rounding on each side
        skip = between(xorig.minimum, touchpad.x.minimum, deviation)
        skip = skip and between(xorig.maximum, touchpad.x.maximum, deviation)
        deviation = 1.5 * touchpad.y.resolution  # 1.5 mm rounding on each side
        skip = skip and between(yorig.minimum, touchpad.y.minimum, deviation)
        skip = skip and between(yorig.maximum, touchpad.y.maximum, deviation)

        if skip:
            print()
            print(
                "Note: Axis ranges within acceptable deviation, skipping min/max override"
            )
            print()

        print()
        print("Suggested hwdb entry:")

        use_dmi = evdev.id["bustype"] not in [0x03, 0x05]  # USB, Bluetooth
        if use_dmi:
            modalias = open("/sys/class/dmi/id/modalias").read().strip()
            print(
                "Note: the dmi modalias match is a guess based on your machine's modalias:"
            )
            print(" ", modalias)
            print(
                "Please verify that this is the most sensible match and adjust if necessary."
            )

        print("-8<--------------------------")
        print("# Laptop model description (e.g. Lenovo X1 Carbon 5th)")
        if use_dmi:
            print("evdev:name:{}:{}*".format(evdev.name, dmi_modalias_match(modalias)))
        else:
            print(
                "evdev:input:b{:04X}v{:04X}p{:04X}*".format(
                    evdev.id["bustype"], evdev.id["vendor"], evdev.id["product"]
                )
            )
        print(
            " EVDEV_ABS_00={}:{}:{}".format(
                touchpad.x.minimum if not skip else "",
                touchpad.x.maximum if not skip else "",
                touchpad.x.resolution,
            )
        )
        print(
            " EVDEV_ABS_01={}:{}:{}".format(
                touchpad.y.minimum if not skip else "",
                touchpad.y.maximum if not skip else "",
                touchpad.y.resolution,
            )
        )
        if evdev.absinfo[libevdev.EV_ABS.ABS_MT_POSITION_X]:
            print(
                " EVDEV_ABS_35={}:{}:{}".format(
                    touchpad.x.minimum if not skip else "",
                    touchpad.x.maximum if not skip else "",
                    touchpad.x.resolution,
                )
            )
            print(
                " EVDEV_ABS_36={}:{}:{}".format(
                    touchpad.y.minimum if not skip else "",
                    touchpad.y.maximum if not skip else "",
                    touchpad.y.resolution,
                )
            )
        print("-8<--------------------------")
        print(
            "Instructions on what to do with this snippet are in /usr/lib/udev/hwdb.d/60-evdev.hwdb"
        )
    except DeviceError as e:
        print("Error: {}".format(e), file=sys.stderr)
        return 1
    except PermissionError:
        print("Unable to open device. Please run me as root", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))
